python/cudf/cudf/tests/test_doctests.py

```diff
@@ -0,0 +1,87 @@
+import doctest
+import inspect
+import os
+
+import numpy as np
+import pytest
+
+import cudf
+
+
+def _name_in_all(parent, name):
+    return name in getattr(parent, "__all__", [])
+
+
+def _is_public_name(parent, name):
+    return not name.startswith("_")
+
+
+def _find_doctests_in_obj(finder, obj, criteria):
+    """Find all doctests in an object.
+
+    Args:
+        finder (doctest.DocTestFinder): The DocTestFinder object to use.
+        obj (module or class): The object to search for docstring examples.
+        criteria (callable): Callable indicating whether to recurse over
```

Comment:
Is this overengineered right now? Do we have a use case for calling this with anything other than a trivial `doctest.DocTestFinder()` or a criteria other than `_name_in_all`? This looks fine, I'm just wary of introducing unnecessary complexity. `_name_in_all` could just be inlined, for instance.

Reply:
This function is called recursively. We need to use a single instance of `doctest.DocTestFinder` that is passed into the first call. The criteria are different for recursing over modules (`_name_in_all`) and classes (`_is_public_name`). I didn't inline them because I thought the names helped explain their purpose, and `_name_in_all` is used multiple times.

Reply:
I improved on this behavior in f4254fdb77.

Reply:
Nice that looks solid to me.